python/ql/src/semmle/python/dataflow/new/internal/DataFlowPublic.qll

```diff
@@ -443,11 +443,94 @@ class BarrierGuard extends GuardNode {
  * - Function parameters
  */
 class LocalSourceNode extends Node {
-  LocalSourceNode() { not simpleLocalFlowStep(_, this) }
+  LocalSourceNode() {
+    not simpleLocalFlowStep+(any(CfgNode n), this) and
+    not this instanceof ModuleVariableNode
+  }
 
   /** Holds if this `LocalSourceNode` can flow to `nodeTo` in one or more local flow steps. */
   cached
-  predicate flowsTo(Node nodeTo) { simpleLocalFlowStep*(this, nodeTo) }
+  predicate flowsTo(Node nodeTo) { Cached::hasLocalSource(nodeTo, this) }
```

Comment:
> Does this provide a better implementation? (Or could we just use `flowsTo` in place of `hasLocalSource`?)

I honestly don't know what you're asking here. Surely we can't use `flowsTo`, as that's the predicate that's being defined. Did you mean renaming `hasLocalSource` to `flowsTo`? That would be the wrong order of arguments.

Reply:
I meant, could we disband `hasLocalSource` and just use the previous implementation of `flowsTo`? And then use `flowsTo` in the places that currently use `hasLocalSource`.

Reply:
We could certainly do that without changing the meaning of the code, but I can't guarantee the same about the performance. In particular, I don't know that the `*` operation is able to use the knowledge that `this` is a `LocalSourceNode` at this point (and so it may end up calculating the entire `simpleLocalFlowStep*` relation before restricting the first argument).

Reply:
Right, that is a valid point. I still think it is a lot easier to read `a.flowsTo(b)` than `hasLocalSource(b, a)` but `flowsTo` is probably not available in the API graph module..

Reply:
... You've lost me again. All uses of `hasLocalSource` are inside `DataFlowPublic` and most of them are inside `Cached`. Everywhere else we use `flowsTo`. :confused: 

I think as part of the general cleanup of the dataflow libraries (mentioned elsewhere) we'll probably want to consider whether `Cached` should be exposed at this point or put somewhere else. Ideally people should want to interact with local flow through `flowsTo`.

Reply:
Ah, you are right, I mixed up the modules. So I _can_ potentially have the readable code, I have submitted suggestions, let me know if that is too optimistic...